### PR TITLE
feat: Ability to invite circles

### PR DESCRIPTION
**Pull request by onny**
```
This commit adds support for adding circles as attendees to a calendar event.
The relationship between the imported group and members will be compliant with
the iCal specification.

A circle with the title "testcircle" will be added as an attendee with iCal
attributes "CUTYPE=GROUP` and uri "mailto:circle+CIRCLEID@CIRCLE_INSTANCE".

Members of the circle will be imported as standard attendees. Each member gets
assigned to the circle group entry by assigning them to the group uri using the
iCal member property: "MEMBER='mailto:circle+CIRCLEID@CIRCLE_INSTANCE'".

Searching for circles is only enabled if the circles app is activated.

Circles added to the list of attendees get imported only once and are not
synced yet. While adding a circle, a notice about this is shown to the user.
Only members of local circles which are local users get imported.

Rendering groups in the frontend is done in a separate PR
https://github.com/nextcloud/calendar/pull/5396
```

image

Closes https://github.com/nextcloud/calendar/issues/2954

Work inspired by @tcitworld PR https://github.com/nextcloud/calendar/pull/4742

### CODE REVIEW

**Comment by st3iny** [comment]
Hey, I just saw this PR and wanted to thank you for tackling this. Could you please add more context to the PR so that we can proceed, e.g what needs to be done and a short description?

It certainly is a nice feature :)

**Comment by onny** [comment]
> Hey, I just saw this PR and wanted to thank you for tackling this. Could you please add more context to the PR so that we can proceed, e.g what needs to be done and a short description?
> 
> It certainly is a nice feature :)

Tank you for your interest. Currently thinking on how to query circle information and the circle members with their email addresses all at once with one request. Propably I'll have to add an additional backend controller / api?

**Comment by onny** [comment]
This is the current state of both PRs combined (rendering groups & invite circles) 🥳 

nextcloud-calendar-invite-circles

**Comment by onny** [comment]
@miaulalala @jancborchardt @nimishavijay @tcitworld @st3iny I tested and simplified the PR a bit more, would be nice to know if there's anything missing :)

**Comment by onny** [comment]
Thank you for reviewing :) Fixed it!

image

**Review by miaulalala — CHANGES_REQUESTED** [review]
Testing this currently.

First thing - the circle membership is unclear with the list of attendees like this:

image

I don't know if it would be possible to have the members displayed as a sub- group. Quick'n'dirty mockup:

image

cc @jancborchardt 

It also sends an invitation email to the one creating the event if they are also a member of the circle. In principle, the ORGANIZER can also be an ATTENDEE, so it's not technically wrong, but it creates awkward visuals as well because the calendar app will show me the accept/decline buttons on an event that I made:

image

Plus I'm a bit concerned about side effects for singificant changes and other stuff if I'm both ORGANIZER and ATTENDEE.

I think it's neccessary to filter out the circle member that is the organizer if that is the case and not set them as an attendee.

Other than that it works flawlessly and I'm super grateful @onny!

**Comment by miaulalala** [comment]
Ah, does https://github.com/nextcloud/calendar/pull/5396 address the comments from above about the display?

**Comment by onny** [comment]
@miaulalala I adjusted the code so that the organizer of the event doesn't get imported from the circle :)

**Comment by miaulalala** [comment]
> @miaulalala I adjusted the code so that the organizer of the event doesn't get imported from the circle :)

I will test it today, thanks a lot!

**Review by miaulalala — COMMENTED** [review]
Looks good! I had a merge conflict that's why it didn't work for me. CI is unhappy, maybe a rebase on main will help.

File: src/models/attendee.js [hunk 1]
```diff
@@ -61,6 +65,8 @@ const mapAttendeePropertyToAttendeeObject = (attendeeProperty) => {
 		role: attendeeProperty.role,
 		rsvp: attendeeProperty.rsvp,
 		uri: attendeeProperty.email,
+		// TODO Expose member attribute in calendar-js
```
Comment (onny) [inline src/models/attendee.js:68]:
```suggestion
		// TODO Expose member attribute in calendar-js
```
easy workaround but propably going to fix this in calendar-js first

**Review by jancborchardt — APPROVED** [review]
Giving my design thumbs up already. :)

**Review by miaulalala — CHANGES_REQUESTED** [review]
Hi onnny, seems CI isn't happy still. 

You will need to run the linter for both PHP (composer cs:fix) and npm (npm run lint:fix). 

Additionally, the exposes internals of the Circles app, that's why statana is complaining here:
`lib/Controller/ContactController.php:212:14: UndefinedClass: Class, interface or enum named OCA\Circles\Api\v1\Circles does not exist `.

The contacts app also retrieves the circles, but via vue: https://github.com/nextcloud/contacts/blob/77771dc2bbca87ea0b108f21192642678eb8dd5d/src/components/AppContent/CircleContent.vue#L112C23-L112C23

I think it's neccessary to move the circle search to the frontend. Sorry about the late change request!

**Comment by onny** [comment]
> Additionally, the exposes internals of the Circles app, that's why statana is complaining here: `lib/Controller/ContactController.php:212:14: UndefinedClass: Class, interface or enum named OCA\Circles\Api\v1\Circles does not exist `.

Maybe I need to figure out how to access the class only if the circles app is enabled and available?

> The contacts app also retrieves the circles, but via vue: https://github.com/nextcloud/contacts/blob/77771dc2bbca87ea0b108f21192642678eb8dd5d/src/components/AppContent/CircleContent.vue#L112C23-L112C23
> 
> I think it's neccessary to move the circle search to the frontend. Sorry about the late change request!

This was my first approach but I guess it's not possible to resolve the user mail addresses via the frontend. My backend controller queries all circle members and then resolves the mail addresses for local users: https://github.com/nextcloud/calendar/blob/7b15357772b98a2e109733e5d3dbde50a2e4afde/lib/Controller/ContactController.php#L239

File: tests/php/unit/Controller/ViewControllerTest.php [hunk 2]
```diff
@@ -122,13 +128,14 @@ public function testIndex(): void {
...
 				['spreed', null, true],
-				['tasks', null, true]
+				['tasks', null, true],
+				['circles', null, false],
 			]);
-		$this->appManager->expects(self::once())
+		$this->appManager->expects(self::exactly(2))
 			->method('getAppVersion')
 			->willReturnMap([
 				['spreed', true, '12.0.0'],
```
Comment (miaulalala) [inline tests/php/unit/Controller/ViewControllerTest.php:141]:
```suggestion
				['spreed', true, '12.0.0'],
				['circles', true, '22.0.0'],

```

File: tests/php/unit/Controller/ViewControllerTest.php [hunk 3]
```diff
@@ -212,13 +220,14 @@ public function testIndexViewFix(string $savedView, string $expectedView): void
...
 				['spreed', null, false],
-				['tasks', null, false]
+				['tasks', null, false],
+				['circles', null, false],
 			]);
-		$this->appManager->expects(self::once())
+		$this->appManager->expects(self::exactly(2))
 			->method('getAppVersion')
 			->willReturnMap([
 				['spreed', true, '11.3.0'],
```
Comment (miaulalala) [inline tests/php/unit/Controller/ViewControllerTest.php:233]:
```suggestion
				['spreed', true, '11.3.0'],
				['circles', true, '22.0.0'],

```

File: psalm.xml [hunk 4]
```diff
@@ -27,6 +27,8 @@
 				<referencedClass name="Doctrine\DBAL\Types\Type" />
 				<referencedClass name="Doctrine\DBAL\Types\Types" />
 				<referencedClass name="OC\Security\CSP\ContentSecurityPolicyNonceManager" />
+				<referencedClass name="OC\App\CompareVersion" />
+				<referencedClass name="OCA\Calendar\Controller\Exception" />
```
Comment (miaulalala) [inline psalm.xml:31]:
```suggestion
				<referencedClass name="OCA\Calendar\Controller\Exception" />
				<referencedClass name="OCA\Circles\Api\v1\Circles" />
```

File: lib/Controller/ViewController.php [hunk 5]
```diff
@@ -117,6 +125,11 @@ public function index():TemplateResponse {
 		$talkApiVersion = version_compare($this->appManager->getAppVersion('spreed'), '12.0.0', '>=') ? 'v4' : 'v1';
 		$tasksEnabled = $this->appManager->isEnabledForUser('tasks');
 
+		$circleVersion = $this->appManager->getAppVersion('circles');
+		$isCirclesEnabled = $this->appManager->isEnabledForUser('circles') === true;
+		// if circles is not installed, we use 0.0.0
+		$isCircleVersionCompatible = $this->compareVersion->isCompatible($circleVersion ? $circleVersion : '0.0.0', '22');
```
Comment (miaulalala) [inline lib/Controller/ViewController.php:131]:
```suggestion
		$isCircleVersionCompatible = $this->compareVersion->isCompatible($circleVersion ?? '0.0.0', '22');
```

**Review by miaulalala — CHANGES_REQUESTED** [review]
Hi onny, I fixed the tests locally, so if you implement the suggestions it should run through. You will also need an additional entry in `psalm.xml`.

Thanks for your hard work - @ChristophWurst and I have agreed since Circles is such a mess and has no public API, it's fine to leave the backend code as is. If you want you can add a comment above the call to the private API in that regard as well that this has been approved but us.

Looking forward to see this in action!!!! :heart_eyes:

File: tests/php/unit/Controller/ViewControllerTest.php [hunk 6]
```diff
@@ -148,7 +155,7 @@ public function testIndex(): void {
 				['show_week_numbers', true],
 				['skip_popover', true],
 				['talk_enabled', true],
-				['talk_api_version', 'v4'],
+				['talk_api_version', 'v1'],
```
Comment (miaulalala) [inline tests/php/unit/Controller/ViewControllerTest.php:158]:
```suggestion
				['talk_api_version', 'v4'],
```

**Review by miaulalala — CHANGES_REQUESTED** [review]

**Comment by onny** [comment]
Thank you @miaulalala and @ChristophWurst for the review. Added the requested changes to the PR :)

**Comment by onny** [comment]
@miaulalala I refactored the code which computes groups and memberships a bit to be more robust and simple. Works quite good for me now :)
Still have a look at the PHP linting errors etc 👀

**Comment by jancborchardt** [comment]
@onny did anything change design-wise? Cause otherwise I would say it is still fine. :)

**Comment by miaulalala** [comment]
@onny can you rebase on the current main? Tests will pass again, then I will take a look at the linter issues.

**Comment by onny** [comment]
> @onny did anything change design-wise? Cause otherwise I would say it is still fine. :)

I refactored the code in `src/components/Editor/Invitees/InviteesList.vue` which computes groups and memberships a bit to be more robust and simple

@miaulalala Rebased the PR

**Comment by miaulalala** [comment]
@ChristophWurst codecov coverage is complaining, am I ok to ignore? I would like to get this merged today 😊

**Comment by ChristophWurst** [comment]
The circles code is probably pain to create test for due to the lack of a stable API so I'll wave this through

**Comment by miaulalala** [comment]
> The circles code is probably pain to create test for due to the lack of a stable API so I'll wave this through

I'll test this one last time during lunch time and then I'll force merge.

Thanks so much for your work and patience @onny !

**Review by jancborchardt — APPROVED** [review]
Still looks good design-wise :)

**Review by miaulalala — APPROVED** [review]
Perfect!

**Pull request merged**